lib/nostrum/cache/guild_cache.ex

```diff
@@ -173,4 +188,172 @@ defmodule Nostrum.Cache.GuildCache do
   @spec select_by!(clauses, selector) :: any | no_return
   def select_by!(clauses, selector),
     do: select_by(clauses, selector) |> Util.bangify_find(clauses, __MODULE__)
+
+  # IMPLEMENTATION
+  @doc false
+  def create(guild) do
+    :ets.insert_new(@table_name, {guild.id, guild})
+  end
+
+  @doc false
+  def update(payload) do
+    [{_id, guild}] = :ets.lookup(@table_name, payload.id)
+    new_guild = Map.merge(guild, payload)
+    true = :ets.update_element(@table_name, payload.id, {2, new_guild})
+  end
+
+  @doc false
+  def delete(guild_id) do
+    # Returns the old guild, if cached
+    case :ets.take(@table_name, guild_id) do
+      [guild] -> guild
+      [] -> nil
+    end
+  end
+
+  @doc false
+  def channel_create(guild_id, channel) do
+    [{_id, guild}] = :ets.lookup(@table_name, guild_id)
+    new_channel = Util.cast(channel, {:struct, Channel})
+    new_channels = Map.put(guild.channels, channel.id, new_channel)
+    new_guild = %{guild | channels: new_channels}
+    true = :ets.update_element(@table_name, guild_id, {2, new_guild})
+    new_channel
+  end
+
+  @doc false
+  def channel_delete(guild_id, channel_id) do
+    [{_id, guild}] = :ets.lookup(@table_name, guild_id)
+    {popped, new_channels} = Map.pop(guild.channels, channel_id)
+    new_guild = %{guild | channels: new_channels}
+    true = :ets.update_element(@table_name, guild_id, {2, new_guild})
+    if popped, do: popped, else: :noop
+  end
+
+  @doc false
+  def channel_update(guild_id, channel) do
+    [{_id, guild}] = :ets.lookup(@table_name, guild_id)
+    {old, new, new_channels} = upsert(guild.channels, channel.id, channel, Channel)
+    new_guild = %{guild | channels: new_channels}
+    true = :ets.update_element(@table_name, guild_id, {2, new_guild})
+    {old, new}
+  end
+
+  @doc false
+  def emoji_update(guild_id, emojis) do
+    [{_id, guild}] = :ets.lookup(@table_name, guild_id)
+    new = %{guild | emojis: emojis}
+    true = :ets.update_element(@table_name, guild_id, {2, new})
+    emojis
```

Comment:
Did we check the return types on these? I took a quick spot check on this one, and for the emoji update event we send both the old and new lists.

Reply:
I'm mostly certain that I checked most return types. 68928905289fd9854e77b732223113093134a72e fixes the emojis.

Reply:
I've rechecked the return types. It looks good to me. I've also added some typespecs.